File: 2020/13/2020-13.py
#Advent of Code 2020 day 13 in micropython
# runs in 216ms on SAMD21g18 (48MHz clock, 32k of RAM)

import time

def read_data(inp):
    sts,rests = inp.split("\n", 1)
    st = int(sts)
    rest = [int(s if s != "x" else "-1") for s in rests.split(",")]
    return st,rest

def part1(data):
    (dtm, all_buses) = data
    buses = [b for b in all_buses if b != -1]
    (bid, delay) = min([(b - (dtm % b), b) for b in buses])
    return bid * delay

def norm(x,b):
    v = x % b
    return v if v >= 0 else v + b

def part2(data):
    (dtm, all_buses) = data
    poly = [(b, norm(-i,b)) for (i,b) in enumerate(all_buses) if b != -1]
    (m,r) = poly[0]
    for (b,i) in poly[1:]:
        while r % b != i:
            r += m
        m *= b
    return r

def main():
    td = read_data("939\n7,13,x,x,59,x,31,19")
    assert part1(td) == 295
    data = read_data(open("input.txt").read())
    print("PART1:", part1(data))

    assert part2(td) == 1068781
    print("PART2:", part2(data))


import gc
st_mem = gc.mem_free() if 'mem_free' in dir(gc) else 0
st_tm = time.monotonic()
main()
end_tm = time.monotonic()
#gc.collect()
end_mem = gc.mem_free() if 'mem_free' in dir(gc) else 0
print("MEM:", end_mem, "USED:",  st_mem-end_mem, "TIME(ms):", (end_tm - st_tm))

File: 2020/13/2020-13.jl

function read_data(inp)
    (ln1, ln2) = split(strip(inp), "\n")
    dtm = parse(Int, ln1)
    buses = [(s == "x" ? -1 : parse(Int, s)) for s in split(ln2, ",")]
    (dtm, buses)
end

tds = """939\n7,13,x,x,59,x,31,19"""

function part1(data)
    (dtm, all_buses) = data
    buses = [b for b in all_buses if b != -1]
    (bid, delay) = minimum([(b - (dtm % b), b) for b in buses])
    bid * delay
end

td = read_data(tds)
@assert part1(td) == 295

data = read_data(read("input.txt", String))
println("PART1: ", part1(data))

norm(x,b) = (x % b) < 0 ? (x % b) + b : x

function part2(data)
    (dtm, all_buses) = data
    poly = [(b, norm(-i+1,b)) for (i,b) in enumerate(all_buses) if b != -1]
    (m,r) = poly[1]
    for (b,i) in poly[2:end]
        while r % b != i
            r += m
        end
        m *= b
    end
    r
end

@assert part2(td) == 1068781
@assert part2(read_data("0\n17,x,13,19")) == 3417
@assert part2(read_data("0\n67,7,59,61")) == 754018
@assert part2(read_data("0\n67,x,7,59,61")) == 779210
@assert part2(read_data("0\n67,7,x,59,61")) == 1261476
@assert part2(read_data("0\n1789,37,47,1889")) == 1202161486

println("PART2: ", part2(data))

